src/Hl7.Fhir.Conformance/Specification/Source/CommonDirectorySource.cs

```diff
@@ -579,20 +579,42 @@ public IEnumerable<ArtifactSummary> ListSummaries()
 
         /// <summary>Find a resource based on its relative or absolute uri.</summary>
         /// <param name="uri">A resource uri.</param>
-        public Resource? ResolveByUri(string uri)
+        public Resource? ResolveByUri(string uri) => TryResolveByUri(uri).Value;
+
+        /// <summary>Find a (conformance) resource based on its canonical uri.</summary>
+        /// <param name="uri">The canonical url of a (conformance) resource.</param>
+        public Resource? ResolveByCanonicalUri(string uri) => TryResolveByCanonicalUri(uri).Value;
+
+        public ResolverResult TryResolveByUri(string uri)
         {
             if (uri == null) throw Error.ArgumentNull(nameof(uri));
             var summary = GetSummaries().ResolveByUri(uri, _inspector);
-            return summary is not null ? loadResourceInternal<Resource>(summary) : null;
+
+            if(summary is null)
+                return ResolverException.ArtifactSummaryNoMatch(uri);
+            
+            var resource = loadResourceInternal<Resource>(summary);
```

Comment:
This function will actually throw already, should we catch that and package in a ResolverResult?